first commit

pyzxxx opened the pull request.

jwu commented on lib/task.js:
```diff
@@ -0,0 +1,286 @@
+}
+
+function getCloneFromType(type, value) {
+  if (type === 'float') {
```
float => number

jwu commented on lib/task.js:
```diff
@@ -0,0 +1,286 @@
+    return vec3.clone(value);
+  } else if (type === 'quat') {
+    return quat.clone(value);
+  }
```
vec4, color3, color4

jwu commented on lib/task.js:
```diff
@@ -0,0 +1,286 @@
+    return quat.clone(value);
+  }
+
+  return null;
```
return value;

jwu commented on lib/task.js:
```diff
@@ -0,0 +1,286 @@
+
+        let to = getValueFromType(type, propInfo.value);
+
+        let from = getCloneFromType(type, target[name]);
```
并不是所有的 from 都是从 target[name] 开始的。我们需要根据参数去判断。例如:

value: [x, y, z] 可以理解为 from 是从当前值，to 是 [x, y, z]
value: [ [x1,y1, z1], [x2, y2, z2] ] 可以理解为 from 是 [x1, y1, z1], to 是 [x2, y2, z2]

jwu commented on lib/task.js:
```diff
@@ -0,0 +1,286 @@
+    return this.reversed ? this.duration - time : time;
+  }
+
+  toggleTaskDirection() {
```
toggle

jwu commented on lib/timeLine.js:
```diff
@@ -0,0 +1,298 @@
+  }
+
+  setCallback(cb) {
+    if (this[cb]) this[cb](this);
```
if (this[cb]) {
  this[cb](this)
}

我们这边书写标准是不要做代码简写。

jwu commented on lib/timeLine.js:
```diff
@@ -0,0 +1,298 @@
+    return this.reversed ? this.duration - time : time;
+  }
+
+  setCallback(cb) {
```
setCallback => invokeCallback

jwu commented on lib/timeLine.js:
```diff
@@ -0,0 +1,298 @@
+    if (this[cb]) this[cb](this);
+  }
+
+  toggleTimeLineDirection() {
```
这个和  toggleTaskDirection 有什么区别？为什么需要两个同样功能的函数?

jwu commented on lib/task.js:
```diff
@@ -0,0 +1,286 @@
+  }
+
+  setCallback(cb) {
+    if (this[cb]) this[cb](this);
```
```
if (this[cb]) {
  this[cb](this)
}
```

我们这边书写标准是不要做代码简写。

jwu commented on lib/task.js:
```diff
@@ -0,0 +1,286 @@
+    }
+  }
+
+  setCallback(cb) {
```
invokeCallback

jwu commented on lib/task.js:
```diff
@@ -0,0 +1,286 @@
+  }
+
+  _setProgress(t) {
+    const tDuration = this.duration;
```
duration

jwu commented on lib/task.js:
```diff
@@ -0,0 +1,286 @@
+
+  _setProgress(t) {
+    const tDuration = this.duration;
+    const tTime = this.adjustTime(t);
```
time

jwu commented on lib/task.js:
```diff
@@ -0,0 +1,286 @@
+
+  tick(t) {
+    this.now = t;
+    if (!this.startTime) this.startTime = this.now;
```
```
if () {
}
```

jwu commented on lib/task.js:
```diff
@@ -0,0 +1,286 @@
+  tick(t) {
+    this.now = t;
+    if (!this.startTime) this.startTime = this.now;
+    const engineTime = this.lastTime + this.now - this.startTime;
```
不需要 const, let 即可。 这里命名用 t 即可不需要 engineTime 这样具体的名字

jwu commented on lib/timeLine.js:
```diff
@@ -0,0 +1,298 @@
+        const key = track.keys[j];
+        if (key.startTime <= tlTime && key.endTime >= tlTime) {
+          //compute elapsed
+          const elapsed = minMaxValue(tlTime - key.startTime - key.delay, 0, key.duration) / key.duration;
```
vmath 里有一个 clamp 可以取代 minMaxValue

jwu commented on lib/task.js:
```diff
@@ -0,0 +1,286 @@
+    for (let i = 0; i < this._animations.length; ++i) {
+      let anim = this._animations[i];
+
+      const elapsed = minMaxValue(time - anim.delay, 0, anim.duration) / anim.duration;
```
vmath 里有一个 clamp 可以取代 minMaxValue

jwu commented on lib/task.js:
```diff
@@ -0,0 +1,286 @@
+  return null;
+}
+
+function getValueFromType(type, value) {
```
parseValue

jwu commented on lib/task.js:
```diff
@@ -0,0 +1,286 @@
+  return null;
+}
+
+function lerpFromType(type) {
```
getLerpFunc

jwu commented on lib/task.js:
```diff
@@ -0,0 +1,401 @@
+          const elapsed = clamp(time - key.startTime - key.delay, 0, key.duration) / key.duration;
+          const eased = isNaN(elapsed) ? 1 : key.easing(elapsed);
+
+          track.lerp(track.target[track.property], key.from, key.to, eased);
```
```js
track.target[track.property] = track.lerp(track.target[track.property], key.from, key.to, eased);
```

jwu commented on lib/task.js:
```diff
@@ -0,0 +1,401 @@
+
+function getLerpFunc(type) {
+  if (type === 'number') {
+    return (to, from, t) => { ((to - from) * t) + from; };
```
和底下的 lerp 对比，参数不符合，调用时会出现错误

pyzxxx committed:
first commit


Second commit


Third commit


third commit


4


5


6


7


vtween v2.0


vtween v2.1


vtween v2.2


vtween v2.3


vtween v2.4

add fixedArray
vtween v2.4

add countIteration()
vtween v2.5

add timeLine
vtween


Merge remote-tracking branch 'origin/master'


Merge remote-tracking branch 'origin/master'
